packages/hashtower.circom/circuits/hashtower-hash-chain.circom

```diff
@@ -0,0 +1,228 @@
+pragma circom 2.1.4;
+
+include "circomlib/circuits/multiplexer.circom";
+include "circomlib/circuits/poseidon.circom";
+include "circomlib/circuits/comparators.circom";
+include "circomlib/circuits/gates.circom";
+
+// Pick in[sel]
+template PickOne(N) {
+    signal input in[N];
+    signal input sel;
+    signal output out;
+
+    component mux = Multiplexer(1, N);
+    for (var i = 0; i < N; i++) { mux.inp[i][0] <== in[i]; }
+    mux.sel <== sel;
+    out <== mux.out[0];
+}
+
+// len 0:          0
+// len 1        in[0]
+// len 2:     H(in[0], in[1])
+// len 3:   H(H(in[0], in[1]), in[2])
+// len 4: H(H(H(in[0], in[1]), in[2]), in[3])
+template HashChain(N) {
+    signal input in[N];
+    signal input len;  // [0...N]
+    signal output out; // one of the N + 1 values
+
+    signal outs[N + 1];
+    outs[0] <== 0;     // empty
+    outs[1] <== in[0]; // no hash
+    for (var i = 2; i < N + 1; i++) {
+        outs[i] <== Poseidon(2)([outs[i - 1], in[i - 1]]);
+    }
+    out <== PickOne(N + 1)(outs, len);
+}
+
+// 0 => 0
+// 1 => 1, 2 => 1, 3 => 1 ...
+template IsNonZero() {
+    signal input in;
+    signal output out <== NOT()(IsZero()(in));
+}
+
+// if in != 1 then abort
+template Must() {
+    signal input in;
+    in === 1;
+}
+
+// if a != b then abort
+template MustEQ() {
```

Comment:
These types of templates make the code harder to read. They also add constraints by defining `a` and `b` (3 constraints vs 1)

Reply:
They are workarounds for the circom compiler.
I hope the can improve readability.

Without them, we might have to create several named temporary components or signals.
They will only be used once, and pollute the scope thereafter.

```circom
pragma circom 2.1.4;

include "circomlib/comparators.circom";

template Example(N) {
    signal input a[N];
    signal input b[N];
    
    signal tmp[N];
    for (var i = 0; i < N; i++) {
        // Error: An anonymous component cannot be used with operator === 
        // 1 === LessThan(8)([a[i], b[i]]);

        tmp[i] <== LessThan(8)([a[i], b[i]]);
        tmp[i] === 1;
        

        // tmp[] will not be used here
    }

    // or here
}
component main = Example(3);

/* INPUT = {
    "a": [ 1, 2, 3 ],
    "b": [ 4, 5, 6 ]
} */
```

Or we can write
```circom
for (var i = 0; i < N; i++) {
    Must()( LessThan(8)([a[i], b[i]]) );
```
to avoid those `tmp`s.